src/riot-rs-threads/src/arch/cortex_m.rs

```diff
@@ -216,53 +212,43 @@ unsafe extern "C" fn PendSV() {
 /// This function is called in PendSV.
 // TODO: make arch independent, or move to arch
 #[no_mangle]
-unsafe fn sched(old_sp: usize) -> usize {
-    let cs = CriticalSection::new();
-    let next_pid;
-
+unsafe fn sched(old_sp: usize) -> u128 {
     loop {
-        {
-            if let Some(pid) = (&*THREADS.as_ptr(cs)).runqueue.get_next() {
-                next_pid = pid;
-                break;
+        if let Some(res) = critical_section::with(|cs| {
+            let threads = &mut *THREADS.as_ptr(cs);
+            if let Some(next_pid) = threads.runqueue.get_next() {
+                let current_high_regs = if let Some(current_pid) = threads.current_pid() {
+                    if next_pid == current_pid {
+                        return Some(0);
+                    }
+                    //println!("current: {} next: {}", current_pid, next_pid);
+                    threads.threads[current_pid as usize].sp = old_sp;
+                    threads.current_thread = Some(next_pid);
+                    threads.threads[current_pid as usize].data.as_ptr()
+                } else {
+                    core::ptr::null()
+                } as usize;
+
+                let next = &threads.threads[next_pid as usize];
+                let next_sp = next.sp as usize;
+                let next_high_regs = next.data.as_ptr() as usize;
+
+                // PendSV expects these three pointers in r0, r1 and r2:
+                // r0 = &next.sp
+                // r1= &current.high_regs
+                // r2= &next.high_regs
+                // On Cortex-M, a u128 as return value is passed in registers r0-r3.
+                // So let's use that.
+                let res: u128 =
+                //  (r0)                     (r1)                        (r2)
+                    (next_sp as u128) |  ((current_high_regs as u128) << 32) | ((next_high_regs as u128) << 64);
+                Some(res)
+            } else {
+                cortex_m::asm::wfi();
```

Comment:
Shouldn't this be outside of the critical section (end of the loop, line 253), so that interrupts are enabled again?

Reply:
It doesn't have to. "wfi" will always stall the CPU until an ISR is *pending*. This code here would then continue, end the critical section, which would then reenable ISRs and jump into them.

Reply:
Ahh okay. I'd still argue that the code is easier to understand if the wfi happens below, but I don't feel strongly about this :).

Reply:
That might actually cause a race. "critical section ends -> isr -> wfi", the wfi could block until the next isr comes in.

Reply:
Ahh good point, I didn't think of that scenario.